### PR TITLE
Step20 add multiple labels to tasks and search for tasks based on labels.

### PR DESCRIPTION
**Pull request by fanfannb**
### Abstract
This PR introduces the ability to add multiple labels to tasks and search for tasks based on labels. These enhancements are intended to improve task categorization and retrieval and make task management more efficient and flexible.

### Features
- Tag Management: Users can assign multiple labels to tasks when creating or editing tasks, but only administrator users can create, edit and delete labels.
- Search by labels: Users can filter tasks by selected labels, improving the ability to quickly find related tasks.

### Motivation
These features were developed to address the need for better task organization within our project management tool. By allowing users to categorize tasks with labels and search by them, we facilitate a more organized and navigable system.

### Implementation Details
- Label Model and Association: Introduced a Label model and a TaskLabel join table to establish a many-to-many relationship between tasks and labels.
- Task Form Updates: Updated the task form for assigning labels to a task.
- Search Functionality: Added a filter option in the task index page that allows users to select labels and filter tasks accordingly.
截屏2024-06-26 17 53 23
截屏2024-06-26 17 51 57
截屏2024-06-26 17 51 46
截屏2024-06-26 17 51 41
截屏2024-06-26 17 51 26
截屏2024-06-26 17 51 19
截屏2024-06-26 17 51 14
截屏2024-06-26 17 54 24

### ステップ20: タスクにラベルをつけられるようにしよう

- タスクに複数のラベルをつけられるようにしてみましょう
- つけたラベルで検索できるようにしてみましょう

### CODE REVIEW

File: myapp/app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -11,13 +11,14 @@ def index
...
-    @tasks = current_user.admin? ? Task.all.order(sort_by => sort_direction) : current_user.tasks.order(sort_by => sort_direction)
+    @tasks = current_user.tasks.order("#{sort_by} #{sort_direction}")
 
     @tasks = @tasks.where('title LIKE ?', "%#{params[:title]}%") if params[:title].present?
-
     @tasks = @tasks.where(status: params[:status]) if params[:status].present?
+    @tasks = @tasks.joins(:labels).where(labels: { id: params[:label_id] }) if params[:label_id].present?
 
     @tasks = @tasks.page(params[:page]).per(12)
+    @labels = Label.all
```
Comment (nagamine-r) [inline myapp/app/controllers/tasks_controller.rb:22]: Can you rewrite the code as common code?
Like `before_action` ?

**Review by nagamine-r — APPROVED** [review]
I think it's mainly OK .

**Review by kanipan — APPROVED** [review]

**Comment by kanipan** [comment]
It's likely that this is a training application, so the design may not have taken performance into much consideration. However, designing applications with performance in mind is very important.

It would be even better if you can design the UI and database in such a way that the response speed remains acceptable even when there are hundreds of millions of records, such as labels, in the database.

I believe you are already aware of this, but please keep it in mind as you develop!

File: myapp/app/controllers/tasks_controller.rb (hunk 1 above)
Comment (fanfannb) [reply myapp/app/controllers/tasks_controller.rb:22]: I changed it.

**Review by nagamine-r — APPROVED** [review]
LGTM

**Comment by fanfannb** [comment]
@kanipan  Thank you for your feedback! I appreciate the reminder and will definitely keep performance considerations in mind as I continue to develop the application.

**Pull request merged**